R/CliUtils.R

```diff
@@ -0,0 +1,197 @@
+  tempconnectionFile <- tempfile(fileext = ".R")
+
+  cli::cli_alert_info("Creating file {tempconnectionFile} for Database credentials - this file will be automaticaly removed when completed and the settings saved securly")
+  on.exit(unlink(tempconnectionFile, force = TRUE))
+
+  rscript <- "# Enter your database connection values here and press save
+# the connection will be tested
+connectionDetails <-
+  DatabaseConnector::createConnectionDetails(
+    dbms = '',
+    server = '',
+    password = '',
+    user = '',
+    port = 0,
+    extraSettings = NULL,
+    connectionString = NULL,
+    pathToDriver = Sys.getenv('DATABASECONNECTOR_JAR_FOLDER')
+  )
+"
+
+  writeLines(rscript, con = tempconnectionFile)
+  connectionValid <- FALSE
+
+  if (!is.null(connectionDetails)) {
+     tryCatch(
+    {
+      source(tempconnectionFile)
+      conn <- DatabaseConnector::connect(connectionDetails)
+      DatabaseConnector::disconnect(conn)
+      connectionValid <- TRUE
+      cli::cli_alert_success("Database Connection Works")
+    },
+      error = function(message) {
+        cli::cli_alert_warning("Database Connection Failed, will require setting manually")
+        cli::cli_alert(message)
+      }
+    )
+  }
+
+  while (!connectionValid) {
+    resp <- utils::menu(c("yes", "no"), title = "Secure creation of connection details required, continue?")
+    if (resp != 1) {
+      cli::cli_abort("Secure database credentials cannot be aquired")
+    }
+    utils::file.edit(tempconnectionFile)
+    # test the connection
+    tryCatch(
+    {
+      source(tempconnectionFile)
+      conn <- DatabaseConnector::connect(connectionDetails)
+      DatabaseConnector::disconnect(conn)
+      connectionValid <- TRUE
+      cli::cli_alert_success("Database Connection Works")
+    },
+      error = function(message) {
+        cli::cli_alert_warning("Database Connection Failed, retrying...")
+        cli::cli_alert(message)
+      }
+    )
+  }
+  unlink(tempconnectionFile, force = TRUE)
+  cli::cli_alert_info("Removed temporary credentials file {tempconnectionFile}")
+
+  cli::cli_h2("Storing values to secure keyring")
+  keyringPassword <- Sys.getenv("STRATEGUS_KEYRING_PASSWORD") # This password is simply to avoid a prompt when creating the keyring
+  # Create the keyring if it does not exist.
+  # If it exists, clear it out so we can re-load the keys
+  allKeyrings <- keyring::keyring_list()
+  if (keyringName %in% allKeyrings$keyring) {
+    if (keyring::keyring_is_locked(keyring = keyringName)) {
+      keyring::keyring_unlock(keyring = keyringName, password = keyringPassword)
+    }
+    # Delete all keys from the keyring so we can delete it
+    message(paste0("Delete existing keyring: ", keyringName))
+    keys <- keyring::key_list(keyring = keyringName)
+    if (nrow(keys) > 0) {
+      for (i in 1:nrow(keys)) {
+        keyring::key_delete(keys$service[i], keyring = keyringName)
+      }
+    }
+    keyring::keyring_delete(keyring = keyringName)
+  }
+
```

Comment:
I think we should change this:   I don't think we should blow away a keyring every time we want to add a new connectionDetailsReference (ie: you should be able to have multiple connectionDetailsReference in 1 keyring).  I think when given a keyring name, we can check to see if it exists or not, and if not...create it...but if so, just leave it alone.